docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,76 @@
+##############################################
+
+Status
+******
+
+**Provisional** 2023-05-23
+
+Context
+*******
+
+Since the initial implementation of the event bus allowed only a single event type to be published to a topic, details like topic name, consumer group name, and consumer name were configured via code. The current implementation of openedx-events does not actually push any events to the underlying implementations like `edx-event-bus-kafka`_ and `edx-event-bus-redis`_. The event-producing application is expected to create a signal handler (since openedx-events subclasses Django signals) to catch the event and push it into the event bus. Some examples of the handlers: `handlers example`_.
+
+.. _handlers example: https://github.com/openedx/edx-platform/blob/27b8d2f68d5dfaf84755e7d7f8dccc97ce3be509/cms/djangoapps/contentstore/signals/handlers.py#L162-L210
+.. _edx-event-bus-kafka: https://github.com/openedx/event-bus-kafka
+.. _edx-event-bus-redis: https://github.com/openedx/event-bus-redis
+
+This ADR aims to propose a solution for configuring the details, like the topic name, consumer group name, etc. via Django settings as well as pushing events to the event bus without requiring the producing application to write additional handlers.
+
+
+Decision
+********
+
+Create a generic signal handler to push events to the event bus. This handler should be attached to or connected to the signals that are enabled in Django settings. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   OPEN_EDX_EVENTS_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'topics': 'xblock-status',
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       },
+       'org.openedx.content_authoring.xblock.published.v1': {
+           'topics': ['xblock-status', 'xblock-published'],
+           'event_key_field': 'xblock_info.usage_key',
+           'event_data_key': 'xblock_info',
+           'enabled': True,
+       },
+   }
+
+This configuration will be read in openedx_events ``apps.OpenedxEventsConfig(AppConfig).ready`` method and a generic signal handler will be connected to the event_types (keys) listed in the configuration after validating its format.
+
+.. code-block:: python
+
+   def ready(self):
+       load_all_signals()
+       config = read_config()
+       # validate the config dictionary and raise errors if any
+       validate_config()
+       for configured_signal in config:
+           connect_or_disconnect_handlers(configured_signal)
+
+The generic handler will again read the configuration and get details for event or signal triggered and push it to the event bus.
+
+.. code-block:: python
+
+   def general_signal_handler(sender, signal, **kwargs):
+       config = read_config(signal)
+       for topic in topics:
+           push_event_to_implementation(signal, config)
+
+Consequences
```

Comment:
We can probably update [OpenEdxPublicSignal](https://github.com/open-craft/openedx-events/blob/4db956585a831d2f40605b9480967fff452d7d0a/openedx_events/content_authoring/signals.py#L23-L28) to include an optional argument called topics and then use the these to generate a default configuration for each signal. This default config can then be updated with the values from users config allowing users to only define the params that they want to change. It would also remove the need for users to define `event_key_field` and `event_data_key` in the config.

I'll update the document on Monday with better explanation of my plan. :)

Reply:
Updated, please have a look at line 26.

Reply:
Just a quick comment here: 
It might be fine to add additional arguments to the events' class definition -- but let's try not to make it that specific since Open edX Events is not necessarily tied to the event bus. Also, I don't see how _topics_ are related to the event definition itself like its data or its current version are. 

If we manage to find an argument justifying why other configurations should belong to the event definition. We could add a: `event_bus_defaults = {}` optional argument, which gives me enough context instead of just _topics_.

Reply:
@mariajgrimaldi: We decided not to move this to the signal definition. See https://github.com/openedx/openedx-events/pull/224#discussion_r1220937831. This can be marked resolved.